### PR TITLE
Use uid's instead of patient/sample ids

### LINKED ISSUE #150: Incorrect result when different studies have same sample ids

### PR DESCRIPTION
**Pull request by kalletlak**
Fix #150

### CODE REVIEW

File: app/scripts/main.js [hunk 1]
```diff
@@ -454,69 +394,80 @@ var iViz = (function(_, $, cbio, QueryByGeneUtil, QueryByGeneTextArea) {
     },
     getCasesMap: function(type) {
       if (type === 'sample') {
-        return data_.groups.group_mapping.sample.patient;
+        return data_.groups.group_mapping.sample_to_patient;
+      }
+      return data_.groups.group_mapping.patient_to_sample;
+    },
+    getcaseUIDs: function(type) {
```
Comment (zhx828) [inline app/scripts/main.js]: Should be named as getCaseUIDs

**Comment by zhx828** [comment]
@kalletlak  in the new changes, if you use rainbow survival plot to draw clinical attribute `CN Cluster` the DFS survival doesn't show anything. When I try to debug, the DFS has been included into `sample` group for some reason. I will continue debugging. And should this go to `master` branch?

**Comment by kalletlak** [comment]
Yes, it is to `master` branch

**Review by yichaoS — APPROVED** [review]

File: app/scripts/vueCore.js [hunk 2]
```diff
@@ -278,10 +269,10 @@
                       self_.hasfilters = true;
                       self_.customfilter.type = group.type;
                       if (radioVal === 'sample') {
-                        self_.customfilter.sampleIds = selectedCaseIds;
+                        self_.customfilter.sampleIds = selectedCaseUIDs;
```
Comment (zhx828) [inline app/scripts/vueCore.js]: In there, what do we use? uid or id? If uid, shouldn't we change the key to sampleUids in customfilter as well? A little confused here.

File: app/scripts/model/dataProxy.js [hunk 3]
```diff
@@ -1155,25 +1165,48 @@ window.DataManagerForIviz = (function($, _) {
...
+          var _patient_uid = 0
           var getSamplesCall = function(cancerStudyId) {
             var def = new $.Deferred();
             window.cbioportal_client.getSamples({
               study_id: [cancerStudyId],
               sample_ids: self.studyCasesMap[cancerStudyId].samples
             }).then(function(data) {
               var sample_to_patient = {};
-              var patientList = {};
+              var sample_to_patient_uid = {};
```
Comment (zhx828) [inline app/scripts/model/dataProxy.js]: This name is a little misleading. I suggest using `sample_uid_to_patient_uid `

File: app/scripts/model/dataProxy.js [hunk 4]
```diff
@@ -621,63 +637,60 @@ window.DataManagerForIviz = (function($, _) {
...
+                  attr_meta: content.util
+                    .sortByClinicalPriority(_.values(_patientAttributes)),
+                  data: _patientUIDData,
+                  has_attr_data: _hasPatientAttrData
+                },
+                sample: {
+                  attr_meta: content.util
+                    .sortByClinicalPriority(_.values(_sampleAttributes)),
+                  data: _sampleUIDData,
+                  has_attr_data: _hasSampleAttrData
```
Comment (zhx828) [inline app/scripts/model/dataProxy.js]: How do we decide the naming convention here?

File: app/scripts/model/dataProxy.js [hunk 5]
```diff
@@ -621,63 +637,60 @@ window.DataManagerForIviz = (function($, _) {
...
+                }
+              };
 
               $.when(self.getCnaFractionData(),
-                self.getMutationCount(),
-                self.getSampleClinicalData(_.keys(_sampleAttributes)),
```
Comment (zhx828) [inline app/scripts/model/dataProxy.js:647]: We no longer caching the full data any more?

File: app/scripts/model/dataProxy.js [hunk 6]
```diff
@@ -1082,37 +1095,34 @@ window.DataManagerForIviz = (function($, _) {
...
+                    cnaSampleIds: [],
+                    allSampleIds: []
+                  };
                   if (_.isArray(self.data.sampleLists.sequenced[cancer_study_id])) {
-                    _sequencedSampleIds = _sequencedSampleIds.concat(self.data.sampleLists.sequenced[cancer_study_id]);
+                    studyCaseList.sequencedSampleIds  = self.data.sampleLists.sequenced[cancer_study_id];
                   }
                   if (_.isArray(self.data.sampleLists.cna[cancer_study_id])) {
-                    _cnaSampleIds = _cnaSampleIds.concat(self.data.sampleLists.cna[cancer_study_id]);
+                    studyCaseList.cnaSampleIds = self.data.sampleLists.cna[cancer_study_id];
```
Comment (zhx828) [inline app/scripts/model/dataProxy.js]: self.data.sampleLists.cna may not be initialized. This will throw exception

File: app/scripts/main.js [hunk 7]
```diff
@@ -454,69 +394,80 @@ var iViz = (function(_, $, cbio, QueryByGeneUtil, QueryByGeneTextArea) {
...
+          'This feature is not available to Virtual Cohort(s) for now!',
           {message_type: 'info'});
       }
     },
     downloadCaseData: function() {
-      var sampleIds_ = vm_.selectedsamples;
+      var sampleUIds_ = vm_.selectedsampleUIDs;
       var attr = {};
+      var self = this;
+      function getAttrVal(attrs, arr) {
```
Comment (zhx828) [inline app/scripts/main.js]: We don't nest function inside function. Maybe put into util?

File: app/scripts/model/dataProxy.js [hunk 8]
```diff
@@ -1082,37 +1095,34 @@ window.DataManagerForIviz = (function($, _) {
...
                   }
                   if (_.isArray(self.data.sampleLists.all[cancer_study_id])) {
-                    _allSampleIds = _allSampleIds.concat(self.data.sampleLists.all[cancer_study_id]);
+                    studyCaseList.allSampleIds = self.data.sampleLists.all[cancer_study_id];
```
Comment (zhx828) [inline app/scripts/model/dataProxy.js:1115]: This needs to be sorted. Please check left panel to make sure all lists are sorted

**Review by zhx828 — CHANGES_REQUESTED** [review]
Done the code review. Before merging, I need to do end-to-end tests. Please wait

File: app/scripts/vueCore.js (hunk 2 above)
Comment (kalletlak) [reply app/scripts/vueCore.js]: Yes, it should be with `Uids`. I'll update it

File: app/scripts/model/dataProxy.js (hunk 4 above)
Comment (kalletlak) [reply app/scripts/model/dataProxy.js]: I'll change it to _`sampleData` and `_patientData`

File: app/scripts/model/dataProxy.js (hunk 6 above)
Comment (kalletlak) [reply app/scripts/model/dataProxy.js]: its initialized at line
https://github.com/cBioPortal/iViz/blob/master/app/scripts/model/dataProxy.js#L969

File: app/scripts/model/dataProxy.js (hunk 8 above)
Comment (kalletlak) [reply app/scripts/model/dataProxy.js:1115]: this new logic is completely different. I am not concatenating samples

File: app/scripts/model/dataProxy.js (hunk 3 above)
Comment (kalletlak) [reply app/scripts/model/dataProxy.js]: Sure, will change that

File: app/scripts/model/dataProxy.js (hunk 5 above)
Comment (kalletlak) [reply app/scripts/model/dataProxy.js:647]: `sampleInitData and patientInitData` clinical data are never used here. And is seems like we are trying to cache data for all clinical attribute with isn't required at this point.

File: app/scripts/main.js (hunk 7 above)
Comment (kalletlak) [reply app/scripts/main.js]: Sure!

**Review by zhx828 — APPROVED** [review]

**Comment by zhx828** [comment]
All public studies passed end-to-end test

**Comment by zhx828** [comment]
@kalletlak The numbers in the header are not changing after selecting few genes.

screen shot 2017-05-17 at 5 45 57 pm

Can't reproduce the issue. Maybe something went wrong when I tested the pull request.

**Pull request merged**